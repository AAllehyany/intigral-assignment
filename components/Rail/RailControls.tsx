import { VariantProps, cva } from 'class-variance-authority'
import ArrowLeftIcon from '../icons/ArrowLeftIcon'
import ArrowRightIcon from '../icons/ArrowRightIcon'

const railControls = cva(
  'absolute flex justify-center items-center top-0 z-2 text-white',
  {
    variants: {
      arrow: {
        left: 'left-0',
        right: 'right-0',
      },
      shape: {
        dark: 'w-16 h-[calc(100%-1.75rem)] bg-black/40 hover:bg-black/50 backdrop-blur-lg',
        transparent: 'w-6 h-full',
      },
    },
    defaultVariants: {
      shape: 'dark',
    },
  },
)

type RailControlsProps = VariantProps<typeof railControls> & {
  handler: () => void
  direction: 'left' | 'right'
  size: 'lg' | 'sm'
}

export default function RailControls({
  handler,
  direction,
  arrow,
  size,
  shape,
}: Readonly<RailControlsProps>) {
  return (
    <button
      data-testid={`rails-controls-${direction}`}
      onClick={handler}
      className={railControls({ arrow, shape })}
    >
      {direction === 'left' && size === 'lg' && (
        <svg
          width="8"
          height="17"
          viewBox="0 0 8 17"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
          className="size-6 fill-current"
        >
          <path opacity="0.6" d="M8 16.5L6.99382e-07 8.5L8 0.5" />
        </svg>
      )}

      {direction === 'left' && size === 'sm' && <ArrowLeftIcon />}

      {direction === 'right' && size === 'lg' && (
        <svg
          width="8"
          height="17"
          viewBox="0 0 8 17"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
          className="size-6 fill-current"
        >
          <path
            opacity="0.6"
            d="M0 16.5L8 8.5L-1.39876e-06 0.5"
            fill="#F2F2F2"
          />
        </svg>
      )}

      {direction === 'right' && size === 'sm' && <ArrowRightIcon />}
    </button>
  )
}
