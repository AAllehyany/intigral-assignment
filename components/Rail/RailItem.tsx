import { ReactNode } from 'react'

type RailItemProps = {
  children: ReactNode
  width: number
}
export default function RailItem({ children, width }: Readonly<RailItemProps>) {
  return (
    <div
      className="w-full shrink-0 grow-0 p-1"
      style={{
        flexBasis: `${width}%`,
        maxWidth: `${width}%`,
      }}
    >
      {children}
    </div>
  )
}
