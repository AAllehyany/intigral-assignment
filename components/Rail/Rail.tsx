'use client'
import RailControls from './RailControls'
import RailItem from './RailItem'
import { ReactNode, useState } from 'react'

type RailProps = {
  rows: number
  children: ReactNode[]
  controls: 'lg' | 'sm'
  shape: 'dark' | 'transparent'
}

export default function Rail({
  rows,
  children,
  controls = 'lg',
  shape = 'dark',
}: Readonly<RailProps>) {
  const rowWidth = 100 / rows

  const maxPages = Math.ceil(children.length / rows)
  const [currentIndex, setCurrentIndex] = useState(0)

  const updateIndex = (value: number) => {
    let calculatedIndex = currentIndex + value
    if (calculatedIndex < 0) calculatedIndex = 0
    let newIndex = calculatedIndex % maxPages

    setCurrentIndex(newIndex)
  }

  return (
    <div className="w-full overflow-x-hidden flex justify-center relative">
      <div
        className="flex w-full items-start transition duration-200"
        style={{
          transform: `translateX(${-100 * currentIndex}%)`,
        }}
      >
        {children.map((child, idx) => (
          <RailItem width={rowWidth} key={idx}>
            {child}
          </RailItem>
        ))}
      </div>

      <RailControls
        size={controls}
        shape={shape}
        arrow="left"
        direction="left"
        handler={() => updateIndex(-1)}
      />
      <RailControls
        size={controls}
        shape={shape}
        arrow="right"
        direction="right"
        handler={() => updateIndex(1)}
      />
    </div>
  )
}
