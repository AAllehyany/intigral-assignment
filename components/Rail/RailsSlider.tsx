import { ReactNode } from 'react'
import RailItem from './RailItem'

type RailsSliderProps = {
  rowWidth: number
  selected: number
  items: ReactNode[]
}

export default function RailsSilder({
  rowWidth,
  selected,
  items,
}: Readonly<RailsSliderProps>) {
  return (
    <div
      className="flex w-full items-start transition duration-200"
      style={{
        transform: `translateX(${-100 * selected}%)`,
      }}
    >
      {items.map((child, idx) => (
        <RailItem width={rowWidth} key={idx}>
          {child}
        </RailItem>
      ))}
    </div>
  )
}
