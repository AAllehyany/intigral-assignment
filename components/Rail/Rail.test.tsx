import { act, fireEvent, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import Rail from './Rail' // Replace with your actual path

it('renders the Rail component', () => {
  const children = [<div>Item 1</div>, <div>Item 2</div>, <div>Item 3</div>]
  const { container } = render(<Rail rows={1}>{children}</Rail>)
  expect(container).toMatchSnapshot()
})

it('correctly navigates to next and previous slides', async () => {
  const mockChildren = Array.from({ length: 5 }, (_, i) => (
    <div key={i}>Item {i + 1}</div>
  ))

  const { container, getByTestId, getByText } = render(
    <Rail rows={1}>{mockChildren}</Rail>,
  )
  const railContainer = getByTestId('rail-container')

  expect(railContainer).toHaveAttribute('data-current', '0')
  const rightArrow = getByTestId('rails-controls-right')
  const leftArrow = getByTestId('rails-controls-left')

  act(() => fireEvent.click(rightArrow))

  expect(railContainer).toHaveAttribute('data-current', '1')
  act(() => fireEvent.click(leftArrow))

  expect(railContainer).toHaveAttribute('data-current', '0')
})

it('loops over when it reaches the end or when it is less than 0', async () => {
  const mockChildren = Array.from({ length: 2 }, (_, i) => (
    <div key={i}>Item {i + 1}</div>
  ))

  const { container, getByTestId, getByText } = render(
    <Rail rows={1}>{mockChildren}</Rail>,
  )
  const railContainer = getByTestId('rail-container')

  expect(railContainer).toHaveAttribute('data-current', '0')
  const rightArrow = getByTestId('rails-controls-right')
  const leftArrow = getByTestId('rails-controls-left')

  act(() => fireEvent.click(leftArrow))

  expect(railContainer).toHaveAttribute('data-current', '0')
  act(() => fireEvent.click(rightArrow))

  expect(railContainer).toHaveAttribute('data-current', '1')

  act(() => fireEvent.click(rightArrow))

  expect(railContainer).toHaveAttribute('data-current', '0')
})
