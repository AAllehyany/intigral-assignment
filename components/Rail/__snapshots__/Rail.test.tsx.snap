// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Rail component 1`] = `
<div>
  <div
    class="w-full overflow-x-hidden flex justify-center relative"
    data-current="0"
    data-testid="rail-container"
  >
    <div
      class="flex w-full items-start transition duration-200"
      style="transform: translateX(0%);"
    >
      <div
        class="w-full shrink-0 grow-0 p-1"
        style="flex-basis: 100%; max-width: 100%;"
      >
        <div>
          Item 1
        </div>
      </div>
      <div
        class="w-full shrink-0 grow-0 p-1"
        style="flex-basis: 100%; max-width: 100%;"
      >
        <div>
          Item 2
        </div>
      </div>
      <div
        class="w-full shrink-0 grow-0 p-1"
        style="flex-basis: 100%; max-width: 100%;"
      >
        <div>
          Item 3
        </div>
      </div>
    </div>
    <button
      class="absolute flex justify-center items-center top-0 z-2 text-white left-0 w-16 h-[calc(100%-1.75rem)] bg-black/40 hover:bg-black/50 backdrop-blur-lg"
      data-testid="rails-controls-left"
    >
      <svg
        class="size-6 fill-current"
        fill="none"
        height="17"
        viewBox="0 0 8 17"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 16.5L6.99382e-07 8.5L8 0.5"
          opacity="0.6"
        />
      </svg>
    </button>
    <button
      class="absolute flex justify-center items-center top-0 z-2 text-white right-0 w-16 h-[calc(100%-1.75rem)] bg-black/40 hover:bg-black/50 backdrop-blur-lg"
      data-testid="rails-controls-right"
    >
      <svg
        class="size-6 fill-current"
        fill="none"
        height="17"
        viewBox="0 0 8 17"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 16.5L8 8.5L-1.39876e-06 0.5"
          fill="#F2F2F2"
          opacity="0.6"
        />
      </svg>
    </button>
  </div>
</div>
`;
